### Назначение ноутбука

В этом ноутбуке реализовано обучение и оценка качества моделей рекомендаций на основе различных подходов:

1. **Baseline 0** — рекомендации по популярности (простая эвристика).
2. **Baseline 1** — модель CatBoost с текстовыми TF-IDF признаками.
3. **MatchingMLP** — нейросетевая модель, обучающаяся на конкатенированных векторах пользователя и товара.

Данные использованы после предварительной фильтрации и формирования негативных пар (rating-based sampling).


In [ ]:
# Импорт необходимых библиотек
import pandas as pd
import numpy as np
from tqdm import tqdm
import sys
import os


disable_tqdm = os.getenv("TQDM_DISABLE", "0") == "1"

### Импорт данных и предобработка

Загружаются:

* `df_train` — обучающая выборка с метками взаимодействия;
* `df_test` — лог реальных взаимодействий пользователей;
* `df_meta` — признаки товаров, включая эмбеддинги CLIP и TF-IDF, числовые и категориальные признаки.

Для использования в разных моделях далее будут созданы отдельные версии датафрейма:

* `df_meta_tfidf` — без CLIP-признаков (только TF-IDF);
* `df_meta_clip` — без TF-IDF (только CLIP + табличные признаки).


In [ ]:
# Загружаем тестовый датасет и обучающую выборку (rating-based sampling)
df_test = pd.read_csv("data/df_test_ground_truth_rating_based.csv")
df_train = pd.read_csv(
    "data/df_train_baseline_0_rating_based.csv",
    na_values=[""],  # исключаем "Unknown"
    keep_default_na=False
)

# Загружаем мета-информацию с признаками TF-IDF и CLIP
df_meta = pd.read_csv(
    "data/amazon_meta_clean.csv",
    na_values=[""],  # исключаем "Unknown"
    keep_default_na=False
)
# Приводим название колонки к общему формату
df_meta.rename(columns={"asin": "item_id"}, inplace=True)
# Удаляем исходные текстовые и визуальные поля, т.к. они не используются напрямую
df_meta.drop(columns=["text_full", "image_main"], inplace=True)

In [3]:
# Для TF-IDF модели: удалим CLIP-фичи
tfidf_cols = [col for col in df_meta.columns if col.startswith("clip_text_") or col.startswith("clip_img_")]
df_meta_tfidf = df_meta.drop(columns=tfidf_cols)

In [4]:
# Для CLIP модели: удалим TF-IDF-фичи
clip_cols = [col for col in df_meta.columns if col.startswith("tfidf_")]
df_meta_clip = df_meta.drop(columns=clip_cols)

## Baseline 0: Popularity-Based Recommender

Наиболее простая стратегия: каждому пользователю рекомендуются самые популярные товары (наиболее часто встречающиеся среди положительных взаимодействий в трейне).

### Преимущества

- Очень быстро считается
- Не требует персонализации или фичей
- Может служить нижней границей качества модели

### Ограничения

- Не учитывает интересы конкретного пользователя
- Невозможно адаптироваться под редкие/новые товары
- Игнорирует временную динамику и мультимодальные признаки

---

Мы сравним этот бейзлайн с более продвинутыми моделями позже: CatBoost, CLIP + текст/изображения и т.д.


In [5]:
# Самые популярные товары (по кол-ву позитивных взаимодействий)
popular_items = (
    df_train[df_train["label"] == 1]["item_id"]
    .value_counts()
    .head(10)
    .index
    .tolist()
)

In [6]:
# Предсказание: каждому пользователю - один и тот же топ-10
predictions = {user: popular_items for user in df_test["user_id"].unique()}

In [ ]:
# Ground truth
# Сопоставляем каждому пользователю множество товаров,
# с которыми он взаимодействовал в тесте
ground_truth = df_test.groupby("user_id")["item_id"].apply(set).to_dict()

## Оценка качества: Precision@K и MAP@K

После генерации рекомендаций для каждого пользователя (`predictions`) мы хотим оценить качество модели с помощью стандартных метрик:

### Precision@K
- Показывает долю релевантных товаров (из ground truth), попавших в top-K рекомендаций
- Рассчитывается для каждого пользователя, затем усредняется

### MAP@K (Mean Average Precision)
- Учитывает не только попадание, но и **позицию** релевантного товара в списке
- Чем выше релевантные товары в списке — тем лучше



In [ ]:
# Функция Precision@K — средняя доля релевантных товаров среди top-K рекомендаций
def precision_at_k(preds, ground_truth, k=10):
    scores = []
    for user, pred_items in preds.items():
        if user not in ground_truth:
            continue
        gt_items = ground_truth[user]
        hits = sum([1 for item in pred_items[:k] if item in gt_items])
        scores.append(hits / k)
    return round(np.mean(scores), 4)

In [ ]:
# Функция average precision для одного пользователя
def apk(pred, actual, k=10):
    if not actual:
        return 0.0
    pred = pred[:k]
    score, num_hits = 0.0, 0.0
    for i, p in enumerate(pred):
        if p in actual and p not in pred[:i]:
            num_hits += 1.0
            score += num_hits / (i + 1.0)
    return score / min(len(actual), k)

In [ ]:
# Средняя average precision по всем пользователям
def map_at_k(preds, ground_truth, k=10):
    return round(
        np.mean([
            apk(preds[u], ground_truth[u], k)
            for u in preds if u in ground_truth
        ]),
        4
    )

In [ ]:
# Оцениваем качество рекомендаций по популярности (Baseline 0)
print("Precision@10:", precision_at_k(predictions, ground_truth, k=10))
print("MAP@10:", map_at_k(predictions, ground_truth, k=10))

Precision@10: 0.0035
MAP@10: 0.0099


## Baseline 1: TF-IDF + PCA + CatBoost

Следующий бейзлайн использует простую, но более информативную стратегию: **машинное обучение на основе текстовых признаков**.

### Идея

- Используем текстовое описание товара (`text_full`) и превращаем его в числовой вектор с помощью **TF-IDF**.
- Чтобы избежать переобучения и снизить размерность — применяем **PCA**.
- Модель **CatBoost** обучается на этих эмбеддингах, предсказывая вероятность релевантности товара для пользователя.

### Преимущества

- Учитывает смысловое содержание описания товара.
- Обеспечивает некоторую степень персонализации.
- Прост в реализации и достаточно эффективен на структурированных данных.

### Ограничения

- Не использует изображение (визуальную составляющую товара).
- Не моделирует поведение пользователя напрямую (работает на item-фичах).
- Возможны потери информации при снижении размерности через PCA.

---

Этот бейзлайн служит первым шагом от простых эвристик к ML-подходам. Далее мы добавим **визуальные эмбеддинги CLIP**, чтобы построить мультимодальную модель.


In [12]:
from catboost import CatBoostClassifier, Pool

In [ ]:
# Загружаем датасет для baseline-модели на TF-IDF фичах
df_train = pd.read_csv(
    "data/df_train_baseline_1_rating_based.csv",
    na_values=[""],  # исключаем "Unknown"
    keep_default_na=False
)

In [ ]:
# Списки признаков: tf-idf фичи, числовые признаки, категориальные признаки
tfidf_cols = [col for col in df_train.columns if col.startswith("tfidf_")]
meta_features = [
    "title_len",
    "title_has_digit",
    "description_text_len",
    "is_top20_brand",
    "has_price",
    "is_top9_category_main"
]
features = (
    tfidf_cols +
    meta_features +
    ["brand", "category_main", "price_clean"]
)
cat_features = ["brand", "category_main"]

In [ ]:
# Пул для catboost
train_pool = Pool(df_train[features], label=df_train["label"], cat_features=cat_features)

In [ ]:
# Инициализируем и обучим модель
model = CatBoostClassifier(
    iterations=1000,
    learning_rate=0.1,
    depth=8,
    l2_leaf_reg=5,
    bootstrap_type="Bayesian",
    eval_metric="AUC",
    random_seed=42,
    verbose=200,
    task_type="GPU"
)

In [17]:
model.fit(train_pool)

Default metric period is 5 because AUC is/are not implemented for GPU


0:	total: 233ms	remaining: 3m 53s
200:	total: 19.4s	remaining: 1m 17s
400:	total: 39.6s	remaining: 59.1s
600:	total: 59.9s	remaining: 39.8s
800:	total: 1m 20s	remaining: 20s
999:	total: 1m 41s	remaining: 0us


### Предсказания и метрики

Для каждого пользователя:

* генерируются топ-10 и топ-50 товаров;
* рассчитываются метрики качества рекомендаций;
* применена фильтрация ранее увиденных товаров.

In [ ]:
# Предсказываем вероятности для всех товаров из df_meta_tfidf
# Сортируем по вероятности и выбираем top-10 и top-50 для каждого пользователя
user_ids = df_test["user_id"].unique()
predictions = {}
predictions_pool = {}

In [ ]:
for user in tqdm(user_ids, file=sys.stdout):
    df_user = df_meta_tfidf.copy()
    df_user["user_id"] = user

    pool = Pool(df_user[features], cat_features=cat_features)
    df_user["pred"] = model.predict_proba(pool)[:, 1]

    # Оригинальный топ-10 (для грязных метрик)
    predictions[user] = (
        df_user.sort_values("pred", ascending=False)["item_id"]
        .head(10)
        .tolist()
    )

    # Расширенный пул топ-50
    predictions_pool[user] = (
        df_user.sort_values("pred", ascending=False)["item_id"]
        .head(50)
        .tolist()
    )

100%|██████████| 38204/38204 [14:46:46<00:00,  1.39s/it]  


In [20]:
ground_truth = df_test.groupby("user_id")["item_id"].apply(set).to_dict()

In [ ]:
# Оцениваем Precision@10 и MAP@10 для модели на TF-IDF по всем товарам
print("Baseline 1 (по всем товарам)")
print("Precision@10:", precision_at_k(predictions, ground_truth, k=10))
print("MAP@10:", map_at_k(predictions, ground_truth, k=10))

Baseline 1 (по всем товарам)
Precision@10: 0.0001
MAP@10: 0.0004


### Фильтрация уже увиденных товаров

Из набора рекомендаций исключаются те товары, с которыми пользователь уже взаимодействовал в трейне — это приближает оценку к реальным условиям использования системы.


In [ ]:
# Удаляем из рекомендаций товары, 
# с которыми пользователь уже взаимодействовал (label = 1)
seen_items = df_train[df_train["label"] == 1].groupby("user_id")["item_id"].apply(set).to_dict()
predictions_filtered = {}

In [ ]:
for user, items in tqdm(predictions_pool.items(), desc="Filtering seen items", file=sys.stdout):
    seen = seen_items.get(user, set())
    filtered = [item for item in items if item not in seen]
    predictions_filtered[user] = filtered[:10]

Filtering seen items: 100%|██████████| 38204/38204 [00:00<00:00, 202866.52it/s]


In [ ]:
# Пересчёт Precision@10 и MAP@10 после удаления просмотренных товаров
print("Baseline 1 (filtered seen items)")
print("Precision@10:", precision_at_k(predictions_filtered, ground_truth, k=10))
print("MAP@10:", map_at_k(predictions_filtered, ground_truth, k=10))

Baseline 1 (filtered seen items)
Precision@10: 0.0
MAP@10: 0.0


# Мэтчинговая MLP модель

### Matching MLP — нейросетевая модель

Формируется обучающая выборка: конкатенируются векторы пользователя (среднее по вектору его положительных товаров) и вектор текущего товара.

Нейросеть `MatchingMLP` обучается отличать положительные пары от отрицательных.



In [ ]:
# Загружаем тренировочную выборку с CLIP-эмбеддингами (текст + изображение)
df_train = pd.read_csv(
    "data/df_train_CLIP_rating_based.csv",
    na_values=[""],  # исключаем "Unknown"
    keep_default_na=False
)

In [ ]:
# Выбираем CLIP-эмбеддинги (текст + изображение)
item_vector_cols = [col for col in df_train.columns if col.startswith("clip_text_") or col.startswith("clip_img_")]
user_ids = df_train["user_id"].unique()

In [ ]:
# Указываем дополнительные табличные признаки, которые добавим к CLIP-вектору
extra_vec_cols = [
    "title_len",
    "description_text_len",
    "is_top20_brand",
    "has_price",
    "is_top9_category_main",
    "price_clean"
]

In [ ]:
# Избавляемся от дубликатов товаров и нормализуем числовые признаки
df_item_extra = df_train.drop_duplicates("item_id")[["item_id"] + extra_vec_cols].copy()

In [ ]:
df_item_extra["price_clean"] = df_item_extra["price_clean"].fillna(-1)

In [30]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df_item_extra[["title_len", "description_text_len", "price_clean"]] = scaler.fit_transform(
    df_item_extra[["title_len", "description_text_len", "price_clean"]]
)

In [ ]:
# Переводим в индекс item_id для удобного объединения
df_item_extra = df_item_extra.set_index("item_id")

In [ ]:
# Извлекаем CLIP-эмбеддинги и объединяем с табличными признаками
clip_vectors  = df_train.drop_duplicates("item_id")[["item_id"] + item_vector_cols].set_index("item_id")

In [ ]:
item_vectors = pd.concat([clip_vectors, df_item_extra], axis=1)

In [34]:
print("extra_vectors shape:", df_item_extra.shape)
print("clip_vectors shape:", clip_vectors.shape)
print("item_vectors shape:", item_vectors.shape)

extra_vectors shape: (53016, 6)
clip_vectors shape: (53016, 200)
item_vectors shape: (53016, 206)


In [ ]:
# Формируем вектор интересов пользователя как среднее по позитивным item-векторам
user_vectors = {}

In [ ]:
for user_id, group in tqdm(df_train[df_train["label"] == 1].groupby("user_id"), desc="User vector aggregation", file=sys.stdout):
    item_ids = group["item_id"].values
    vectors = item_vectors.loc[item_ids].values
    user_vectors[user_id] = np.mean(vectors, axis=0)

User vector aggregation: 100%|██████████| 236983/236983 [01:44<00:00, 2261.55it/s]


In [ ]:
# Строим датасет: конкатенируем векторы пользователя и товара
X = []
y = []

In [ ]:
for row in tqdm(df_train.itertuples(), total=len(df_train), desc="Building training pairs", file=sys.stdout):
    item_id = row.item_id
    user_id = row.user_id
    label = row.label

    if user_id not in user_vectors or item_id not in item_vectors.index:
        continue

    user_vec = user_vectors[user_id]
    item_vec = item_vectors.loc[item_id].values

    concat_vec = np.concatenate([user_vec, item_vec])
    X.append(concat_vec)
    y.append(label)

Building training pairs: 100%|██████████| 531220/531220 [01:34<00:00, 5649.48it/s]


In [ ]:
# Преобразуем списки в массивы для дальнейшей подачи в PyTorch
X = np.array(X)
y = np.array(y)
print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (489315, 412)
y shape: (489315,)


In [ ]:
# Импортируем PyTorch и определяем, использовать ли GPU
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader, random_split

In [41]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Device:", device)

Device: cuda


In [ ]:
# Определяем Dataset
class MatchingDataset(Dataset):
    def __init__(self, X, y):
        self.X = torch.tensor(X, dtype=torch.float32)
        self.y = torch.tensor(y, dtype=torch.float32)

    def __len__(self):
        return len(self.X)

    def __getitem__(self, idx):
        return self.X[idx], self.y[idx]

In [ ]:
# Создаём датасет и делим его на обучающую и валидационную части
full_dataset = MatchingDataset(X, y)

In [ ]:
train_size = int(0.9 * len(full_dataset))
val_size = len(full_dataset) - train_size
train_dataset, val_dataset = random_split(full_dataset, [train_size, val_size])

In [ ]:
# Оборачиваем датасеты в DataLoader для удобной подачи батчами
train_loader = DataLoader(train_dataset, batch_size=2048, shuffle=True)
val_loader = DataLoader(val_dataset, batch_size=2048)

In [ ]:
# Определяем MLP модель
class MatchingMLP(nn.Module):
    def __init__(self, input_dim=400):
        super().__init__()
        self.model = nn.Sequential(
            nn.Linear(input_dim, 256),
            nn.ReLU(),
            nn.Dropout(0.2),
            nn.Linear(256, 64),
            nn.ReLU(),
            nn.Linear(64, 1),
            nn.Sigmoid()
        )

    def forward(self, x):
        return self.model(x)

In [ ]:
# Создаём экземпляр модели
model = MatchingMLP(input_dim=X.shape[1]).to(device)

In [ ]:
# Используем бинарную кросс-энтропию в качестве функции потерь
# и Adam как оптимизатор с learning rate 0.001
criterion = nn.BCELoss()
optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)

In [ ]:
from sklearn.metrics import accuracy_score
# Обучаем модель в течение 10 эпох, отслеживая loss и accuracy на валидации
num_epochs = 10

In [ ]:
for epoch in range(num_epochs):
    model.train()
    train_losses = []

    for X_batch, y_batch in tqdm(train_loader, desc="Training", file=sys.stdout):
        X_batch, y_batch = X_batch.to(device), y_batch.to(device).view(-1, 1)

        preds = model(X_batch)
        loss = criterion(preds, y_batch)

        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        train_losses.append(loss.item())

    # Валидация
    model.eval()
    val_losses = []
    all_preds = []
    all_targets = []

    for X_batch, y_batch in tqdm(val_loader, desc="Validating", file=sys.stdout):
        X_batch, y_batch = X_batch.to(device), y_batch.to(device).view(-1, 1)
        with torch.no_grad():
            preds = model(X_batch)
            loss = criterion(preds, y_batch)
        val_losses.append(loss.item())
        all_preds.extend(preds.cpu().numpy())
        all_targets.extend(y_batch.cpu().numpy())

    # Бинаризуем предсказания, считаем accuracy и выводим метрики за эпоху
    all_preds_bin = (np.array(all_preds) >= 0.5).astype(int)
    val_acc = accuracy_score(all_targets, all_preds_bin)

    tqdm.write(
        f"Epoch {epoch+1}/{num_epochs} | "
        f"Train Loss: {np.mean(train_losses):.4f} | "
        f"Val Loss: {np.mean(val_losses):.4f} | "
        f"Val Accuracy: {val_acc:.4f}"
    )


Epoch 1/10


Epoch 1/10 | Train Loss: 0.2844 | Val Loss: 0.2337 | Val Accuracy: 0.9193

Epoch 2/10


Epoch 2/10 | Train Loss: 0.2157 | Val Loss: 0.1918 | Val Accuracy: 0.9333

Epoch 3/10


Epoch 3/10 | Train Loss: 0.1909 | Val Loss: 0.1752 | Val Accuracy: 0.9373

Epoch 4/10


Epoch 4/10 | Train Loss: 0.1758 | Val Loss: 0.1620 | Val Accuracy: 0.9435

Epoch 5/10


Epoch 5/10 | Train Loss: 0.1605 | Val Loss: 0.1478 | Val Accuracy: 0.9479

Epoch 6/10


Epoch 6/10 | Train Loss: 0.1448 | Val Loss: 0.1379 | Val Accuracy: 0.9518

Epoch 7/10


Epoch 7/10 | Train Loss: 0.1302 | Val Loss: 0.1288 | Val Accuracy: 0.9549

Epoch 8/10


Epoch 8/10 | Train Loss: 0.1162 | Val Loss: 0.1128 | Val Accuracy: 0.9619

Epoch 9/10


Epoch 9/10 | Train Loss: 0.1032 | Val Loss: 0.1016 | Val Accuracy: 0.9652

Epoch 10/10


Epoch 10/10 | Train Loss: 0.0927 | Val Loss: 0.0933 | Val Accuracy: 0.9682


In [ ]:
# Проверяем размер и разнообразие тестового сета перед инференсом
print(f"Test shape: {df_test.shape}, unique users: {df_test['user_id'].nunique()}")

Test shape: (44324, 5), unique users: 38204


In [ ]:
# Создаём словарь ground truth: реальные товары, 
# с которыми взаимодействовал пользователь
ground_truth = df_test.groupby("user_id")["item_id"].apply(set).to_dict()

In [ ]:
# Переводим модель в режим инференса
model.eval()

MatchingMLP(
  (model): Sequential(
    (0): Linear(in_features=412, out_features=256, bias=True)
    (1): ReLU()
    (2): Dropout(p=0.2, inplace=False)
    (3): Linear(in_features=256, out_features=64, bias=True)
    (4): ReLU()
    (5): Linear(in_features=64, out_features=1, bias=True)
    (6): Sigmoid()
  )
)

In [ ]:
# Подготавливаем список кандидатов и их эмбеддинги для инференса
candidate_items = item_vectors.index.tolist()
candidate_vectors = item_vectors.values
candidate_vectors.shape

(53016, 206)

### Предсказания и метрики

Для каждого пользователя:

* генерируются топ-10 и топ-50 товаров;
* рассчитываются метрики качества рекомендаций;
* применена фильтрация ранее увиденных товаров.

In [ ]:
# Для каждого пользователя генерируем top-N рекомендаций
# Сортировка кандидатов по вероятности релевантности (score)
top_k = 10
predictions = {}
predictions_pool = {}

In [ ]:
for user_id in tqdm(ground_truth.keys(), desc="Generating top-N", file=sys.stdout):
    if user_id not in user_vectors:
        continue

    user_vec = user_vectors[user_id]
    user_vec_batch = np.tile(user_vec, (candidate_vectors.shape[0], 1))
    concat = np.hstack([user_vec_batch, candidate_vectors])

    with torch.no_grad():
        scores = model(torch.tensor(concat, dtype=torch.float32).to(device)).cpu().numpy().flatten()

    sorted_items = np.array(candidate_items)[np.argsort(scores)[::-1]]
    predictions[user_id] = sorted_items[:10].tolist()
    predictions_pool[user_id] = sorted_items[:50].tolist()

Generating top-N: 100%|██████████| 38204/38204 [2:26:18<00:00,  4.35it/s]  


In [ ]:
# Считаем метрики на всём списке рекомендаций
print("Matching MLP (top-10)")
print("Precision@10:", precision_at_k(predictions, ground_truth))
print("MAP@10:", map_at_k(predictions, ground_truth))

Matching MLP (top-10)
Precision@10: 0.01
MAP@10: 0.0644


In [ ]:
# Выясняем с какими товарами user уже взаимодействовал
seen_items = df_train[df_train["label"] == 1].groupby("user_id")["item_id"].apply(set).to_dict()

In [ ]:
# Очищаем предсказания от уже увиденных товаров
predictions_filtered = {}
for user, items in tqdm(predictions_pool.items(), desc="Filtering seen items", file=sys.stdout):
    seen = seen_items.get(user, set())
    filtered = [item for item in items if item not in seen]
    predictions_filtered[user] = filtered[:10]

Filtering seen items: 100%|██████████| 33860/33860 [00:00<00:00, 126756.62it/s]


In [ ]:
# Считаем финальные метрики качества рекомендаций после фильтрации seen-items
print("Matching MLP (filtered seen items)")
print("Precision@10:", precision_at_k(predictions_filtered, ground_truth))
print("MAP@10:", map_at_k(predictions_filtered, ground_truth))

Matching MLP (filtered seen items)
Precision@10: 0.003
MAP@10: 0.0131


### Сравнение моделей по метрикам качества (n-pair sampling)

| Модель                  | Тестовый сценарий   | Precision\@10 | MAP\@10    |
| ----------------------- | ------------------- | ------------- | ---------- |
| **Baseline 0**          | по всем товарам     | 0.0035    | 0.0099     |
| **Baseline 1 (TF-IDF)** | по всем товарам     | 0.0001        | 0.0004     |
|                         | filtered seen items | 0.0       | 0.0     |
| **Matching MLP (CLIP)** | по всем товарам     | **0.01**        | **0.0644** |
|                         | filtered seen items | 0.003        | 0.0131    |

### Выводы:

* **Matching MLP** вновь демонстрирует **лучшие результаты** среди всех моделей:

  * **MAP\@10**: 0.0644 по полному пулу товаров;
  * Это в **6+ раз выше**, чем у бейзлайна по популярности.
* **Baseline 0** остаётся простой, но достаточно сильной отправной точкой.
* **Baseline 1 (TF-IDF)** **значительно уступает** даже популярности — особенно при фильтрации уже виденных товаров. Это подтверждает, что **TF-IDF признаки слабо отражают пользовательские предпочтения** в условиях рейтинг-бейзда.
* **Фильтрация seen-items** приводит к ожидаемому снижению метрик — особенно у Matching MLP, но итоговый **MAP\@10 всё ещё почти в 1.5 раза выше, чем у Baseline 0**.
* **Вывод:** мультимодальный подход с эмбеддингами CLIP остаётся наиболее перспективным, даже при использовании только реальных рейтингов без негативных сэмплов.


